app/models/remote_authenticatable.rb

```diff
@@ -1,92 +1,44 @@
+require 'auth/authentication_backend'
+
 module Devise
   module Models
     module RemoteAuthenticatable
       extend ActiveSupport::Concern
+      include AuthenticationBackend
 
-      #
-      # Here you do the request to the external webservice
-      #
-      # If the authentication is successful you should return
-      # a resource instance
-      #
-      # If the authentication fails you should return false
-      #
       def remote_authentication(params)
-        if params.key?(:cognito_session_id)
-          resp = respond_to_challenge(params)
-        else
-          resp = initiate_auth(params[:email], params[:password])
-        end
-        return process_response(resp, params) if resp.present?
+        resp = authentication_flow(params)
 
-        raise StandardError("No Response Back from AWS to process")
-      end
-
-      def process_response(resp, params)
-        if resp.challenge_name.present?
+        if resp.class == ChallengeResponse
```

Comment:
in terms of abstraction, it would be better not to need to know about the class name (`ChallengeResponse`) in this layer but all more like `resp.challenge?` or similar as part of the object? ... just a thought ;)

Reply:
I would expect almost all Auth Providers supporting 2FA to have some form of Challenge Response.  I didn't view this as a name specific to Cognito and I believe Challenge Response is the standard term for this style of flow.

In terms of interfaces I would standardise on a specific set of classes which we provide.  If we switched to Auth0 or Keycloak I would still expect to either get an AuthenticatedResponse or a ChallengeResponse.

Reply:
I don't dispute that that different providers will have some sort of the challenge. My point was that the consumer of that module shouldn't need to know about class names. If this becomes a gem, users would need to change their code if decide to rename the internal classes/working...

Reply:
FINE! 😝😁

OK thats a good point... I shall add a challenge_type field to the classes and we'll check for that in our code.  In my own head remote_authenticatable would also form part of the gem.

Reply:
cool, it doesn't have to be anything specific, but just a way to expose whether the response is for a challenge or auth.. and checking by different means than a class name ;)